syncnotification.go

```diff
@@ -34,99 +30,397 @@ func (lw *LibWallet) spvSyncNotificationCallbacks(loadedWallet *wallet.Wallet) *
 		RescanProgress:               generalNotifications.RescanProgress,
 		RescanFinished:               generalNotifications.RescanFinished,
 		PeerDisconnected: func(peerCount int32, addr string) {
-			for _, syncResponse := range lw.syncProgressListeners {
-				syncResponse.OnPeerDisconnected(peerCount)
-			}
+			lw.handlePeerCountUpdate(peerCount)
 		},
 		PeerConnected: func(peerCount int32, addr string) {
-			for _, syncResponse := range lw.syncProgressListeners {
-				syncResponse.OnPeerConnected(peerCount)
-			}
+			lw.handlePeerCountUpdate(peerCount)
 		},
 	}
 }
 
-func (lw *LibWallet) generalSyncNotificationCallbacks(loadedWallet *wallet.Wallet) *chain.Notifications {
+func (lw *LibWallet) generalSyncNotificationCallbacks() *chain.Notifications {
 	return &chain.Notifications{
-		Synced: func(sync bool) {
-			// begin indexing transactions after sync is completed,
-			// syncProgressListeners.OnSynced() will be invoked after transactions are indexed
-			lw.IndexTransactions(-1, -1, func() {
-				for _, syncResponse := range lw.syncProgressListeners {
-					syncResponse.OnSynced(sync)
-				}
-			})
-		},
-		FetchMissingCFiltersStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(0, 0, SyncStateStart)
-			}
-		},
-		FetchMissingCFiltersProgress: func(missingCFitlersStart, missingCFitlersEnd int32) {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(missingCFitlersStart, missingCFitlersEnd, SyncStateProgress)
-			}
-		},
-		FetchMissingCFiltersFinished: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(0, 0, SyncStateFinish)
-			}
-		},
-		FetchHeadersStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(0, 0, SyncStateStart)
-			}
-		},
-		FetchHeadersProgress: func(fetchedHeadersCount int32, lastHeaderTime int64) {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(fetchedHeadersCount, lastHeaderTime, SyncStateProgress)
-			}
-		},
-		FetchHeadersFinished: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(0, 0, SyncStateFinish)
-			}
-		},
-		DiscoverAddressesStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnDiscoveredAddresses(SyncStateStart)
-			}
-		},
-		DiscoverAddressesFinished: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnDiscoveredAddresses(SyncStateFinish)
-			}
+		Synced:                       lw.synced,
```

Comment:
Also re-arrange these functions in this `*chain.Notifications` object in the order that we expect they'd be called. I think moving  only `Synced` to the bottom (after `RescanFinished`) would achieve that.

Also, re-arrange the equivalent `lw.---` functions to match the order they appear here. `lw. fetchHeadersStarted` should appear first before `lw.fetchHeadersProgress` and so on.

The `lw.handlePeerCountUpdate` function will appear before `lw. fetchHeadersStarted` since it is called first in the scheme of things.

The last set of functions would be `lw.notifySyncCanceled`, `lw.notifySyncError` and `lw.synced`